### PR TITLE
QueryBuilder now supports `Traversable` objects for `IN` clauses.

### PR DESCRIPTION
**Pull request by SamMousa**

### CODE REVIEW

**Comment by SilverFire** [comment]
Changelog line, please

**Comment by SamMousa** [comment]
You'd think I'd learn... :8ball:

**Comment by SamMousa** [comment]
Implements #10954

File: framework/db/QueryBuilder.php [hunk 1]
```diff
@@ -1112,7 +1113,7 @@ public function buildInCondition($operator, $operands, &$params)
             return $this->buildSubqueryInCondition($operator, $column, $values, $params);
         }
 
-        $values = (array) $values;
+        $values = ArrayHelper::toArray($values);
```
Comment (cebe) [inline framework/db/QueryBuilder.php]: should set recursive parameter to false to avoid complete iteration over the values if it is already an array.

File: framework/db/QueryBuilder.php [hunk 2]
```diff
@@ -1202,13 +1203,13 @@ protected function buildCompositeInCondition($operator, $columns, $values, &$par
             }
             $vss[] = '(' . implode(', ', $vs) . ')';
         }
+
+        $sqlColumns = [];
         foreach ($columns as $i => $column) {
-            if (strpos($column, '(') === false) {
-                $columns[$i] = $this->db->quoteColumnName($column);
-            }
+            $sqlColumns[] = strpos($column, '(') === false ? $this->db->quoteColumnName($column) : $column;
```
Comment (cebe) [inline framework/db/QueryBuilder.php]: why did you change this?

Comment (SamMousa) [inline framework/db/QueryBuilder.php]: Ah, my intention was to not call `ArrayHelper::toArray()` and only iterate the values once regardless of it being an array or `\Traversable`.

Comment (cebe) [inline framework/db/QueryBuilder.php]: Oh, now I see. makes sense.

**Comment by SamMousa** [comment]
Still working on some improvements; while reading the code I found that `buildCompositeInCondition` does not support db expressions; is there a reason for that?

**Comment by cebe** [comment]
I do not think there is a specific reason, `Expression` support has been added partially over time to more and more methods.

**Comment by SamMousa** [comment]
Ok, my implementation is ready so feedback is welcome :)

**Comment by SamMousa** [comment]
Okay some tests that I've added are failing and I'm not sure what the best way to fix them is.

Basically `sqlite` and `mssql` drivers do not support composite in conditions like:
```` where (id, name) in ((1, "Bob"), (2, "Jane"))````

Previously there was no test case for this, but I've added one. The new test fails because there is 1 dataprovider that is used for all query builders, but not all query builders have the same expected value.

Should I instead move these tests to a different test function and not use the dataprovider?

**Comment by SilverFire** [comment]
I remember about your PR, but do not have enough time to review it right now

**Comment by SilverFire** [comment]
Could you updated from master?

**Comment by SamMousa** [comment]
Yes I will, but regardless tests will fail until I know how to solve that ;-)

**Comment by samdark** [comment]
Yes, these test should be driver-specific.

**Comment by SilverFire** [comment]
Fixed tests and merged. Thank you @SamMousa :beers: 
https://github.com/yiisoft/yii2/commit/dcf42e35cc114a8a4652140c0c6d198eaeac2a66
